### ISSUE TITLE
refactor stringbuf/cstr class

### ISSUE THREAD

**Issue by rgerhards** [rsyslog/rsyslog#1033]
There is to be a lot of legacy in that code and related inconsistencies. Among others, I wonder why we really need these two string representations. But unfortunately my memory doesn't tell me any longer why we wanted this in the first place... My line of thought is if we reserve one extra byte of memory in the regular string buffer, we should be able to create a standard C string representation without any need for double buffers. Also, I think the partial handling of NUL bytes stems back to CEE, which initially wanted to support it, but later dropped. IMHO we can remove all of that support without any problems... I think this list of issues is not conclusive, the review will bring up more. All in all, a great simplification should be possible.

For some more background, see also #1028.

**Comment by rgerhards** [comment]
Note: NUL byte support is indeed **not** needed. For example, the very important getMSG() function in msg.c assumes a classical C string. So if NUL bytes are present, they would be handled incorrectly and lead to message truncation. It is the input's core to make sure this requirement is met.

If we find that inputs do not handle this consistently, we could add a guard to MsgSetRawMsg(), however, this has potentially large runtime implications.

**Comment by rgerhards** [comment]
Review of the input side reveals that NUL bytes are handled at the parser stage. The parser must either handle control character sanitization itself (and escape at a minimum the NUL bytes) or it agrees to use the default sanitizer, which escapes NUL bytes. The latter is what currently happens with all parsers. A sample can be found here: https://github.com/rsyslog/rsyslog/blob/master/tools/pmrfc5424.c#L63

In conclusion, we do not need to handle embedded NUL bytes inside strings being processed after the parser stage. Note that parsing happens on ```pMsg->pszRawMsg```, which is a classic C string buffer and so ```stringbuf.c``` is not involved before the parsing stage (this is intentional).
